### PR TITLE
Move optimizer inside the DualProblem

### PR DESCRIPTION
**Pull request by guilhermebodin**

### CODE REVIEW

**Comment by guilhermebodin** [comment]
This closes #20

File: src/MOI_wrapper.jl [hunk 1]
```diff
@@ -13,15 +15,16 @@ const SS = Union{MOI.EqualTo{Float64}, MOI.GreaterThan{Float64}, MOI.LessThan{Fl
...
 mutable struct DualOptimizer{OT <: MOI.ModelLike} <: MOI.AbstractOptimizer
-    dual_problem::Union{Nothing, DualProblem}
-    dual_optimizer::OT
-    dual_optimizer_idx_map::Union{Nothing, MOIU.IndexMap}
+    dual_problem::DualProblem
 
-    function DualOptimizer{OT}(dual_optimizer::OT) where {OT <: MOI.ModelLike}
-        return new(nothing, dual_optimizer, nothing)
+    function DualOptimizer{OT}(dual_problem::DualProblem) where {OT <: MOI.ModelLike}
+        return new(dual_problem)
```
Comment (matbesancon) [inline src/MOI_wrapper.jl]:
```suggestion
        return new{OT}(dual_problem)
```

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -13,15 +15,16 @@ const SS = Union{MOI.EqualTo{Float64}, MOI.GreaterThan{Float64}, MOI.LessThan{Fl
...
     end
     function DualOptimizer(dual_optimizer::OT) where {OT <: MOI.ModelLike}
```
Comment (matbesancon) [inline src/MOI_wrapper.jl]: These two constructors can be external constructors

File: src/MOI_wrapper.jl [hunk 3]
```diff
@@ -110,38 +101,25 @@ function get_vis_dual_problem(optimizer::DualOptimizer, ci::CI)
...
-    dual_optimizer_vi = Vector{VI}(undef, length(vis_dual_problem))
-    for (i, vi) in enumerate(vis_dual_problem)
-        dual_optimizer_vi[i] = optimizer.dual_optimizer_idx_map.varmap[vi]
-    end
-    return dual_optimizer_vi
-end
-
 function MOI.get(optimizer::DualOptimizer, ::MOI.SolverName)
-    return "Dual model with "*MOI.get(optimizer.dual_optimizer, MOI.SolverName())*" attached"
+    return "Dual model with "*MOI.get(optimizer.dual_problem.dual_model, MOI.SolverName())*" attached"
```
Comment (matbesancon) [inline src/MOI_wrapper.jl]:
```suggestion
    return "Dual model with "*MOI.get(optimizer.dual_problem.dual_model, MOI.SolverName()) * " attached"
```

File: src/MOI_wrapper.jl [hunk 4]
```diff
@@ -200,17 +176,17 @@ end
...
 
 function MOI.get(optimizer::DualOptimizer, ::MOI.DualStatus)
-    return MOI.get(optimizer.dual_optimizer, MOI.PrimalStatus())
+    return MOI.get(optimizer.dual_problem.dual_model, MOI.PrimalStatus())
 end
 
 function MOI.get(optimizer::DualOptimizer, ::MOI.ResultCount)
-    return MOI.get(optimizer.dual_optimizer, MOI.ResultCount())
+    return MOI.get(optimizer.dual_problem.dual_model, MOI.ResultCount())
 end
```
Comment (matbesancon) [inline src/MOI_wrapper.jl]:
```suggestion
end

```

File: src/dualize.jl [hunk 5]
```diff
@@ -6,42 +6,48 @@ export dualize
 Dualize the model
 """
 function dualize(primal_model::MOI.ModelLike; dual_names::DualNames = DualNames("", ""))
-    return dualize(primal_model, dual_names, Float64)
+    # Crates an empty dual problem
```
Comment (matbesancon) [inline src/dualize.jl]:
```suggestion
    # Creates an empty dual problem
```

File: src/dualize.jl [hunk 6]
```diff
@@ -6,42 +6,48 @@ export dualize
...
 end
 
-function dualize(primal_model::MOI.ModelLike, dual_names::DualNames, T::DataType)
+function dualize(primal_model::MOI.ModelLike, dual_problem::DualProblem{T}; 
+                 dual_names::DualNames = DualNames("", "")) where T
+    # Dualize with the optimizer already attached
+    return dualize(primal_model, dual_problem, dual_names)
+end
+
+function dualize(primal_model::MOI.ModelLike, dual_problem::DualProblem{T}, dual_names::DualNames) where T
```
Comment (matbesancon) [inline src/dualize.jl]: You don't seem to be using `T` in the function
```suggestion
function dualize(primal_model::MOI.ModelLike, dual_problem::DualProblem, dual_names::DualNames)
```

File: src/dualize.jl [hunk 7]
```diff
@@ -6,42 +6,48 @@ export dualize
...
     # Add dual equality constraint and get the link dictionary
-    add_dual_equality_constraints(dual_model, primal_model,
-                                  primal_dual_map, dual_names,
+    add_dual_equality_constraints(dual_problem.dual_model, primal_model,
+                                  dual_problem.primal_dual_map, dual_names,
                                   primal_objective, con_types)
 
-    return DualProblem(dual_model, primal_dual_map)
+    return dual_problem
 end
```
Comment (matbesancon) [inline src/dualize.jl]:
```suggestion
end

```

File: src/structures.jl [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    function DualProblem{T}() where T
+        return new(DualizableModel{T}(), PrimalDualMap{T}())
+    end
+    function DualProblem{T}(dual_optimizer::OT) where {T, OT <: MOI.ModelLike}
+        return new(dual_optimizer, PrimalDualMap{T}())
+    end
+    function DualProblem(dual_optimizer::OT) where {OT <: MOI.ModelLike}
+        return DualProblem{Float64}(MOIU.CachingOptimizer(DualizableModel{Float64}(), dual_optimizer))
+    end
+end
```
Comment (matbesancon) [inline src/structures.jl]:
```suggestion
end

```

File: test/Tests/test_structures.jl [hunk 9]
```diff
@@ -0,0 +1,6 @@
+@testset "structures" begin
+    primal_dual_map = Dualization.PrimalDualMap{Float64}()
+    @test Dualization.is_empty(primal_dual_map)
+    push!(primal_dual_map.primal_var_dual_con, VI(1) => CI{SVF, MOI.EqualTo}(1))
+    @test !Dualization.is_empty(primal_dual_map)
+end
```
Comment (matbesancon) [inline test/Tests/test_structures.jl]:
```suggestion
end

```

**Comment by guilhermebodin** [comment]
@matbesancon thank you for the review!

**Pull request merged**